include/graphene-box2d.h

```diff
@@ -0,0 +1,122 @@
+                                                                     const graphene_point_t *max);
+GRAPHENE_AVAILABLE_IN_1_12
+graphene_box2d_t *        graphene_box2d_init_from_points           (graphene_box2d_t       *box,
+                                                                     unsigned int            n_points,
+                                                                     const graphene_point_t *points);
+GRAPHENE_AVAILABLE_IN_1_12
+graphene_box2d_t *        graphene_box2d_init_from_vectors          (graphene_box2d_t       *box,
+                                                                     unsigned int            n_vectors,
+                                                                     const graphene_vec2_t  *vectors);
+GRAPHENE_AVAILABLE_IN_1_12
+graphene_box2d_t *        graphene_box2d_init_from_box              (graphene_box2d_t       *box,
+                                                                     const graphene_box2d_t *src);
+GRAPHENE_AVAILABLE_IN_1_12
+graphene_box2d_t *        graphene_box2d_init_from_vec2             (graphene_box2d_t       *box,
+                                                                     const graphene_vec2_t  *min,
+                                                                     const graphene_vec2_t  *max);
+
+GRAPHENE_AVAILABLE_IN_1_12
+void                      graphene_box2d_expand                     (const graphene_box2d_t *box,
+                                                                     const graphene_point_t *point,
+                                                                     graphene_box2d_t       *res);
+GRAPHENE_AVAILABLE_IN_1_12
+void                      graphene_box2d_expand_vec2                (const graphene_box2d_t *box,
+                                                                     const graphene_vec2_t  *vec,
+                                                                     graphene_box2d_t       *res);
+GRAPHENE_AVAILABLE_IN_1_12
+void                      graphene_box2d_expand_scalar              (const graphene_box2d_t *box,
+                                                                     float                   scalar,
+                                                                     graphene_box2d_t       *res);
+GRAPHENE_AVAILABLE_IN_1_12
+void                      graphene_box2d_union                      (const graphene_box2d_t *a,
+                                                                     const graphene_box2d_t *b,
+                                                                     graphene_box2d_t       *res);
+GRAPHENE_AVAILABLE_IN_1_12
+bool                      graphene_box2d_intersection               (const graphene_box2d_t *a,
+                                                                     const graphene_box2d_t *b,
+                                                                     graphene_box2d_t       *res);
```

Comment:
intersection checking is the one absolutely most critical operation we are doing when rendering in GTK.

In fact, in certain workloads (think a GtkListbox in a scrolledwindow, like `gtk4-demo --run=listbox`) intersecting alone can take >50% of time spend in `gsk_renderer_render()` (according to sysprof).

So having an (inline?) function that just does an intersection check and returns `true/false` would be very neat. Something like `bool graphene_box2d_intersects (a, b)`

Reply:
I started adding a static inline `intersects()` operator, but accessing private fields in the header file gets gnarly. I'm open to that, but it needs some more finesse than just giving access to the contents of the structure to everyone.

Reply:
With `graphene_rect_t` we have basically inlined everything, see https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gsk/gskrectprivate.h for what we do there.

And it's a massive win for performance, because the compiler can not just avoid function calls, but it can inline the entire code, which means it can reorder instructions any way it likes.  

We measured significant speedups from that in the critical path (text rendering), which runs millions of times per second when benchmarking and simple graphene function calls like `graphene_rect_scale()` took multiple percentage points of total performance according to sysprof.

Here's roughly that example with GTK git main:

```
diff --git a/gsk/gpu/gskgpunodeprocessor.c b/gsk/gpu/gskgpunodeprocessor.c
index 7467ee72e1..2cb70d8a7c 100644
--- a/gsk/gpu/gskgpunodeprocessor.c
+++ b/gsk/gpu/gskgpunodeprocessor.c
@@ -3104,14 +3104,18 @@ gsk_gpu_node_processor_add_glyph_node (GskGpuNodeProcessor *self,
                                                  &glyph_bounds,
                                                  &glyph_offset);
 
-      glyph_tex_rect = GRAPHENE_RECT_INIT (-glyph_bounds.origin.x / scale,
-                                           -glyph_bounds.origin.y / scale,
-                                           gsk_gpu_image_get_width (image) / scale,
-                                           gsk_gpu_image_get_height (image) / scale);
-      glyph_bounds = GRAPHENE_RECT_INIT (0,
-                                         0,
-                                         glyph_bounds.size.width / scale,
-                                         glyph_bounds.size.height / scale);
+      graphene_rect_scale (&GRAPHENE_RECT_INIT (-glyph_bounds.origin.x,
+                                                -glyph_bounds.origin.y,
+                                                gsk_gpu_image_get_width (image),
+                                                gsk_gpu_image_get_height (image)),
+                           1 / scale, 1 / scale,
+                           &glyph_tex_rect);
+      graphene_rect_scale (&GRAPHENE_RECT_INIT (0,
+                                                0,
+                                                glyph_bounds.size.width,
+                                                glyph_bounds.size.height),
+                           1 / scale, 1 / scale,
+                           &glyph_bounds);
       glyph_origin = GRAPHENE_POINT_INIT (glyph_origin.x - glyph_offset.x / scale,
                                           glyph_origin.y - glyph_offset.y / scale);
 ```

With that patch applied, `GDK_DEBUG=no-vsync GSK_DEBUG=full-redraw build/demos/widget-factory/gtk4-widget-factory` of a `-O3` GTK build drops from ~1850fps to ~1780fps - which is somewhere around 3-5%. And here's what sysprof says:

![image](https://github.com/user-attachments/assets/624df36f-509b-464f-af14-ba0ec4e9f5a3)

Reply:
I'm not against inlining things—after all, that's why all `graphene_simd4f_t` functions are inline instead of using the pixman approach of a run time check for the CPU architecture. The main issue is that the higher order vectorised types in Graphene have private fields from the perspective of the caller, which makes writing static inlines harder. Unlike `graphene_rect_t`, which has public fields for a reason, other types like `graphene_vec[234]_t` and `graphene_box2d_t` have their public fields mangled when not compiling Graphene.

In any case, I've added a `graphene_box2d_intersects()` which inlines *most* of the intersection; it still has to call `graphene_box2d_get_minmax()` to get the minimum and maximum vertices; shaving more calls and getting direct access to the SIMD type gets complicated unless I completely abandon the field name mangling.